File: python/evalkit_plugin/src/evalkit_plugin/runtime.py
from __future__ import annotations

import json
import sys
from dataclasses import dataclass
from typing import Any, Callable

PLUGIN_PROTOCOL_VERSION = "1"


@dataclass(frozen=True)
class PluginSpec:
    kind: str
    name: str
    version: str
    capabilities: tuple[str, ...]


@dataclass(frozen=True)
class PluginError(Exception):
    code: str
    message: str
    details: Any = None


def acquisition_plugin(
    name: str,
    *,
    version: str = "0.1.0",
    capabilities: tuple[str, ...] | list[str] = (),
) -> Callable[[Callable[[str], str]], Callable[[str], str]]:
    return _decorate_plugin("acquisition", name, version, capabilities)


def scorer_plugin(
    name: str,
    *,
    version: str = "0.1.0",
    capabilities: tuple[str, ...] | list[str] = (),
) -> Callable[[Callable[..., dict[str, Any]]], Callable[..., dict[str, Any]]]:
    return _decorate_plugin("scorer", name, version, capabilities)


def _decorate_plugin(
    kind: str,
    name: str,
    version: str,
    capabilities: tuple[str, ...] | list[str],
) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
    spec = PluginSpec(
        kind=kind,
        name=name,
        version=version,
        capabilities=tuple(capabilities),
    )

    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
        setattr(func, "__evalkit_plugin_spec__", spec)
        return func

    return decorator


def run_plugin(plugin: Callable[..., Any]) -> None:
    spec = getattr(plugin, "__evalkit_plugin_spec__", None)
    if spec is None:
        raise TypeError("plugin must be decorated with acquisition_plugin or scorer_plugin")

    request = _read_request()
    _write_json(
        {
            "kind": spec.kind,
            "name": spec.name,
            "version": spec.version,
            "schema_version": PLUGIN_PROTOCOL_VERSION,
            "capabilities": list(spec.capabilities),
        }
    )

    try:
        if spec.kind == "acquisition":
            response = {"output": plugin(request["input"])}
        elif spec.kind == "scorer":
            response = {
                "score": plugin(
                    request["input"],
                    request["output"],
                    request.get("reference"),
                    request.get("run_id"),
                    request.get("sample_id"),
                    request["trial_index"],
                    request.get("metadata", {}),
                )
            }
        else:
            raise RuntimeError(f"unsupported plugin kind: {spec.kind}")
    except PluginError as err:
        response = {
            "error": {
                "code": err.code,
                "message": err.message,
                "details": err.details if err.details is not None else {},
            }
        }

    _write_json(response)


def _read_request() -> dict[str, Any]:
    line = sys.stdin.readline()
    if not line:
        raise RuntimeError("expected one JSON request line on stdin")

    payload = json.loads(line)
    if not isinstance(payload, dict):
        raise TypeError("plugin request must be a JSON object")
    return payload


def _write_json(payload: dict[str, Any]) -> None:
    sys.stdout.write(json.dumps(payload, separators=(",", ":")))
    sys.stdout.write("\n")
    sys.stdout.flush()

File: typescript/evalkit_plugin/src/runtime.ts
export const PLUGIN_PROTOCOL_VERSION = "2";

const pluginSpecSymbol = Symbol.for("evalkit.plugin.spec");

export type PluginKind = "source" | "scorer";
export type PluginCapability = string;

export interface PluginSpec {
  kind: PluginKind;
  name: string;
  version: string;
  capabilities: PluginCapability[];
}

export interface ScorerRequestMetadata {
  [key: string]: unknown;
}

type PluginWithSpec<F extends (...args: never[]) => unknown> = F & {
  [pluginSpecSymbol]: PluginSpec;
};

export type SourcePlugin = PluginWithSpec<
  (input: unknown) => unknown | Promise<unknown>
>;

export type ScorerPlugin = PluginWithSpec<
  (
    input: unknown,
    output: unknown,
    reference: unknown,
    runId: string | undefined,
    sampleId: string | undefined,
    trialIndex: number,
    metadata: ScorerRequestMetadata,
  ) => unknown | Promise<unknown>
>;

export class PluginError extends Error {
  readonly code: string;
  readonly details: unknown;

  constructor(code: string, message: string, details: unknown = {}) {
    super(message);
    this.name = "PluginError";
    this.code = code;
    this.details = details;
  }
}

export function sourcePlugin(
  name: string,
  options: {
    version?: string;
    capabilities?: PluginCapability[];
  } = {},
): <F extends (input: unknown) => unknown | Promise<unknown>>(plugin: F) => SourcePlugin {
  return decoratePlugin("source", name, options);
}

export function scorerPlugin(
  name: string,
  options: {
    version?: string;
    capabilities?: PluginCapability[];
  } = {},
): <
  F extends (
    input: unknown,
    output: unknown,
    reference: unknown,
    runId: string | undefined,
    sampleId: string | undefined,
    trialIndex: number,
    metadata: ScorerRequestMetadata,
  ) => unknown | Promise<unknown>,
>(
  plugin: F,
) => ScorerPlugin {
  return decoratePlugin("scorer", name, options);
}

export async function runPlugin(plugin: SourcePlugin | ScorerPlugin): Promise<void> {
  const spec = plugin[pluginSpecSymbol];

  if (spec === undefined) {
    throw new TypeError(
      "plugin must be decorated with sourcePlugin(...) or scorerPlugin(...)",
    );
  }

  const request = await readRequest();
  writeJson({
    kind: spec.kind,
    name: spec.name,
    version: spec.version,
    schema_version: PLUGIN_PROTOCOL_VERSION,
    capabilities: spec.capabilities,
  });

  try {
    if (spec.kind === "source") {
      const output = await (plugin as SourcePlugin)(request.input);
      writeJson({ output });
      return;
    }

    const score = await (plugin as ScorerPlugin)(
      request.input,
      request.output,
      request.reference,
      asOptionalString(request.run_id),
      asOptionalString(request.sample_id),
      asTrialIndex(request.trial_index),
      asMetadata(request.metadata),
    );
    writeJson({ score });
  } catch (error) {
    if (error instanceof PluginError) {
      writeJson({
        error: {
          code: error.code,
          message: error.message,
          details: error.details ?? {},
        },
      });
      return;
    }

    throw error;
  }
}

function decoratePlugin<K extends PluginKind>(
  kind: K,
  name: string,
  options: {
    version?: string;
    capabilities?: PluginCapability[];
  },
) {
  const spec: PluginSpec = {
    kind,
    name,
    version: options.version ?? "0.1.0",
    capabilities: options.capabilities ?? [],
  };

  return <F extends (...args: never[]) => unknown>(plugin: F): PluginWithSpec<F> => {
    const decorated = plugin as PluginWithSpec<F>;
    decorated[pluginSpecSymbol] = spec;
    return decorated;
  };
}

async function readRequest(): Promise<Record<string, unknown>> {
  const chunks: Buffer[] = [];

  for await (const chunk of process.stdin) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }

  const input = Buffer.concat(chunks).toString("utf8").trim();
  if (input.length === 0) {
    throw new Error("expected one JSON request line on stdin");
  }

  const payload = JSON.parse(input) as unknown;
  if (payload === null || typeof payload !== "object" || Array.isArray(payload)) {
    throw new TypeError("plugin request must be a JSON object");
  }

  return payload as Record<string, unknown>;
}

function writeJson(payload: Record<string, unknown>): void {
  process.stdout.write(`${JSON.stringify(payload)}\n`);
}

function asOptionalString(value: unknown): string | undefined {
  return typeof value === "string" ? value : undefined;
}

function asTrialIndex(value: unknown): number {
  return typeof value === "number" && Number.isInteger(value) ? value : 0;
}

function asMetadata(value: unknown): ScorerRequestMetadata {
  if (value === null || value === undefined || typeof value !== "object" || Array.isArray(value)) {
    return {};
  }

  return value as ScorerRequestMetadata;
}
